### PR TITLE
feat(expr): distributed make timestamptz

### PR DESCRIPTION
**Pull request by KeXiangWang**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Support distributed `make_timestamptz`. 
`CapturedContext` is introduced for pass session info like timezone. 

There will be two part where we need to utilized the captured context: 
1. When executing the executors, the contexted is used to finish the caluculation.
2. When building exectuors, we may need to call the `execute` RPC in `task_service` to create a task on other CN as data source. At this time, we need to the capture these values and pass these values to the CN throught the `execute` RPC.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: proto/batch_plan.proto [hunk 1]
```diff
@@ -359,3 +359,7 @@ message PlanFragment {
   PlanNode root = 1;
   ExchangeInfo exchange_info = 2;
 }
+
+message CapturedContext {
+  optional string time_zone = 1;
```
Comment (fuyufjh) [inline proto/batch_plan.proto:364]: As it's optional, in which situation it can be missed?

File: proto/batch_plan.proto [hunk 2]
```diff
@@ -359,3 +359,7 @@ message PlanFragment {
   PlanNode root = 1;
   ExchangeInfo exchange_info = 2;
 }
+
+message CapturedContext {
```
Comment (fuyufjh) [inline proto/batch_plan.proto:363]: Here the `CapturedContext` message is defined in `batch_plan.proto`, while in `src/expr/core/src/captured_context.rs`, it's said to be "For all execution mode".

So shall we move that to somewhere else like `expr.proto` or `compute.proto` or `common.proto`?

Btw, please add some comments for it.

File: src/frontend/src/scheduler/distributed/stage.rs [hunk 3]
```diff
@@ -622,11 +632,16 @@ impl StageRunner {
     }
 
     async fn schedule_tasks_for_all(&mut self, shutdown_rx: ShutdownToken) -> SchedulerResult<()> {
+        let captured_context = CapturedContext {
+            // TODO(kexiang): We go through the plan to make sure only the neccessay context is captured.
```
Comment (fuyufjh) [inline src/frontend/src/scheduler/distributed/stage.rs:636]: This sounds error-prone because it needs the programmer to *manually* ensure a context variable is provided if a function is used. I prefer to capture all variables no matter being used or not. 

After serialization, the context should be small. If we do want to save some bytes, we may omit the default values.

**Review by fuyufjh — COMMENTED** [review]
Thanks for the work! Generally LGTM.

**Review by BugenZhao — COMMENTED** [review]
From my understanding, the contexts were all previously defined in the `frontend` crate because...

- We are sure they're evaluated in the frontend (local mode), as they're mostly admin functions.
- The context types are only available in frontend crate.

According to those, it seems reasonable to extract the `TIME_ZONE` context out of `frontend` and into `expr` crate. As a result, we have to find all paths to provide the context before evaluating. However, I'm considering whether there's really a strong reason for not inlining the timezone into the expression itself (as a constant argument). This is how we handle `at_timezone` and `now` before, which is really easy to implement and not invasive at all.

**Comment by BugenZhao** [comment]
> inlining the timezone into the expression itself

IIUC, simply adding a match arm of `ExprType::MakeTimestampTz` with 6 arguments and rewrite it to 7 arguments would work. 🤔

File: src/frontend/src/scheduler/distributed/stage.rs (hunk 3 above)
Comment (KeXiangWang) [reply src/frontend/src/scheduler/distributed/stage.rs:636]: Agree.

File: src/batch/src/rpc/service/task_service.rs [hunk 4]
```diff
@@ -119,12 +121,14 @@ impl TaskService for BatchServiceImpl {
...
             epoch,
             tracing_context,
+            captured_context,
         } = req.into_inner();
 
         let task_id = task_id.expect("no task id found");
         let plan = plan.expect("no plan found").clone();
         let epoch = epoch.expect("no epoch found");
         let tracing_context = TracingContext::from_protobuf(&tracing_context);
+        let captured_context = captured_context.expect("no captured context found");
```
[the remark this answers is not in the document]
Comment (KeXiangWang) [reply src/batch/src/rpc/service/task_service.rs:131]: `time_zone` is optional. Here it seems to be optional because Protobuf always compile a user-defined message to an `Option<>`. But actually, the field is always filled except for testing.

File: proto/batch_plan.proto (hunk 2 above)
Comment (KeXiangWang) [reply proto/batch_plan.proto:363]: I'm about to move it `stream_plan` as we also need this for streaming execution mode. 

For naming, what about `SessionContext`? As this struct is mainly used to capture various sessional values for various execution modes.

Comment (liurenjie1024) [reply proto/batch_plan.proto:363]: I would prefer to name it `ExecutionContext` since it's about one query execution. It may include things even more limited to one execution

**Comment by KeXiangWang** [comment]
> IIUC, simply adding a match arm of `ExprType::MakeTimestampTz` with 6 arguments and rewrite it to 7 arguments would work. 🤔

Yes, current rewrite machism is friendly to expr part but cause some problems on frontend's logics. Thanks to @xiangjinwu for helping find the following issues caused by current rewrite mechanism:

https://github.com/risingwavelabs/risingwave/pull/8415
https://github.com/risingwavelabs/risingwave/pull/8403
The backend constructs its own expr and bypasses the frontend rewrite.

https://github.com/risingwavelabs/risingwave/issues/7175#issuecomment-1381314486
https://github.com/risingwavelabs/risingwave/pull/13640 
Some parts' rewrite is missed.

https://github.com/risingwavelabs/risingwave/pull/10086
https://github.com/risingwavelabs/risingwave/pull/9580
https://github.com/risingwavelabs/risingwave/pull/12633 
The frontend rewrite can be finished in different stage. And sometimes the rewrite can cause conflict with the logic of `is_now_offset` and `WatermarkAnalyzer`.

https://github.com/risingwavelabs/risingwave/pull/12633#discussion_r1347313875
https://github.com/risingwavelabs/risingwave/pull/12640
Invalid expr caused by not rewrite.

https://github.com/risingwavelabs/risingwave/pull/10410
https://github.com/risingwavelabs/risingwave/pull/13201
Another example of wrong rewrite logics. 

In summary, current rewrite mechanism introduce challenges on frontend rewrite. It's easy to miss or incorrectly perform the rewrite. That's why I'm now trying to redesign the implementation of timezone, starting with `make_timestamptz` func.

This PR's implementation is somehow tricky but avoids the error-prone rewrite. 

We can discuss which is better.

File: proto/batch_plan.proto (hunk 2 above)
Comment (KeXiangWang) [reply proto/batch_plan.proto:363]: `ExecutionContext` is good, but it conflicts with [an exsiting struct](https://github.com/risingwavelabs/risingwave/blob/2ddab6bc806040455c139823ed75cbcea3ef143f/src/frontend/src/scheduler/mod.rs#L47) in frontend codes. I'm concerning it will cause confusion.

Comment (liurenjie1024) [reply proto/batch_plan.proto:363]: Or `ExecutionContextInfo`? In fact I expect the existing struct should contains such info.

Comment (KeXiangWang) [reply proto/batch_plan.proto]: `Context`and `Info` together seems to be a little redundant. Maybe `CapturedExecutionContext`?

Comment (liurenjie1024) [reply proto/batch_plan.proto]: LGTM

**Review by fuyufjh — APPROVED** [review]
LGTM

**Review by liurenjie1024 — APPROVED** [review]

File: src/batch/src/task/task_execution.rs [hunk 5]
```diff
@@ -433,14 +436,17 @@ impl<C: BatchTaskContext> BatchTaskExecution<C> {
...
         );
 
-        let exec = ExecutorBuilder::new(
-            self.plan.root.as_ref().unwrap(),
-            &self.task_id,
-            self.context.clone(),
-            self.epoch.clone(),
-            self.shutdown_rx.clone(),
+        let exec = captured_execution_context_scope!(
+            captured_execution_context.clone(),
```
Comment (xiangjinwu) [inline src/batch/src/task/task_execution.rs:440]: > As a result, we have to find all paths to provide the context before evaluating. ...(the frontend inline rewriting approach) is really easy to implement and not invasive at all.

We either need to find all paths to provide the context before evaluating, or find all paths to inline the context before evaluating. The primary goal of these `make_timestamptz` PRs are not to support this new function for users, but to be the pioneer and see if the new approach is better than existing one.

@KeXiangWang How did you figure out these places to provide context in this PR? If some of them slip through, how easy are they exposed as a runtime error? My impression was that there are less places and easier to surface compared to the expr rewrite approach (recent addition #13724). But this can be wrong and simply due to we had more time to receive bugs on the rewrite approach.

I also agree inlining is less invasive, to the extend that it is solely in frontend. But if approach A requires awareness in 10 submodules in 1 module, while approach B requires awareness in 2 submodules in 3 parent modules each, it may also be acceptable to consider A as more invasive because 10 > 2 * 3. Again these numbers are made up arbitrarily and personally I feel the difference between 10 and 6 is not large enough to justify it.

**Review by xiangjinwu — APPROVED** [review]
Thanks for establishing the new route!

File: proto/batch_plan.proto (hunk 2 above)
Comment (xiangjinwu) [reply proto/batch_plan.proto]: > I'm about to move it stream_plan as we also need this for streaming execution mode.

Why not `plan_common.proto`?

Comment (KeXiangWang) [reply proto/batch_plan.proto]: 😂It‘s actually not used in `batch_plan.proto`, so `stream_plan.proto` is enough. But `plan_common.proto` make more sense. I'll move it to `plan_common.proto`.

File: src/batch/src/task/task_execution.rs (hunk 5 above)
Comment (KeXiangWang) [reply src/batch/src/task/task_execution.rs:440]: I went through the codes in frontend and run some test examples to help verify whether all of them are found. 

The good thing is when there's an expression been evaluated without context, it will throw an error like "context TIME_ZONE not found" and the error is shown in psql console, making it easy to recognize. For streaming mode, it will generate one log with the same error message and when we read from the MV, the result of that row is missed.

Another benefit is that we only need to find all the paths (to provide the context before evaluating) once. Once it's finished correctly with `make_timestamptz`, unless we have new paths, we don't need to care it in following the implementation of other expressions.

File: src/expr/macro/src/lib.rs [hunk 6]
```diff
@@ -645,3 +649,37 @@ pub fn capture_context(attr: TokenStream, item: TokenStream) -> TokenStream {
...
+/// Add scope to provide the captured context variables for the closure.
+/// For now, we add this when building an Executor and when an Exetutor is run.
+#[proc_macro]
+pub fn captured_execution_context_scope(input: TokenStream) -> TokenStream {
+    fn inner(input: TokenStream) -> Result<TokenStream2> {
+        let CapturedExecutionContextScopeInput { context, closure } = syn::parse(input)?;
+
+        let ctx = quote! { let ctx = #context; };
+        let mut body = quote! { #closure };
+        let fields = vec![format_ident!("time_zone")];
```
[the remark this answers is not in the document]
Comment (KeXiangWang) [reply src/expr/macro/src/lib.rs:662]: Let leave it to the streaming mode PR.
